### PR TITLE
Support masked marks in the clientside_mark directive

### PR DESCRIPTION
**Pull request by gozzy**
... to preserve existing mark bits when adding Squid connmarks.

No functionality changes expected until the /mask is configured.

The masking implementation is based on iptables --set-mark rather than
--set-xmark principles. TODO: Add support for the latter if needed.

### CODE REVIEW

File: src/cache_cf.cc [hunk 1]
```diff
@@ -1542,9 +1544,12 @@ static void
 dump_acl_nfmark(StoreEntry * entry, const char *name, acl_nfmark * head)
 {
     for (acl_nfmark *l = head; l; l = l->next) {
-        if (l->nfmark > 0)
+        if (l->nfmark > 0) {
             storeAppendPrintf(entry, "%s 0x%02X", name, l->nfmark);
-        else
+
+            if (l->nfmask != 0xffffffff)
+                storeAppendPrintf(entry, "/%02X", l->nfmask);
```
Comment (rousskov) [inline src/cache_cf.cc]: Similarly to parsing discussed in another change request, we ought to reuse the printing code/approach used in Acl::ConnMark::dump() instead of half-duplicating that code and introducing inconsistencies. AFAICT, that reuse would be easy once the ConnMarkConfig class (described in that other change request) is introduced. The existing `std::ostream` printing method (and possibly more) would become available for dumping clientside_mark values.

File: src/client_side_request.cc [hunk 2]
```diff
@@ -1794,9 +1794,14 @@ ClientHttpRequest::doCallouts()
             ACLFilledChecklist ch(NULL, request, NULL);
             ch.src_addr = request->client_addr;
             ch.my_addr = request->my_addr;
-            nfmark_t mark = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch);
-            if (mark)
+            acl_nfmark *l = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch);
+            if (l) {
```
Comment (rousskov) [inline src/client_side_request.cc]: Please merge the above two lines together to avoid exposing the rest of the code to (a possibly nil) `l`:

```
if (const auto mc = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch))
```

File: src/cf.data.pre [hunk 3]
```diff
@@ -2570,7 +2570,15 @@ DOC_START
 	Allows you to apply a Netfilter mark value to packets being transmitted
 	on the client-side, based on an ACL.
 
-	clientside_mark mark-value [!]aclname ...
+	clientside_mark mark[/mask] [!]aclname ...
+
+	Mark and mask are unsigned integers (hex, octal, or decimal).
+	Mask can be used if a mark set by other agents before squid (e.g., iptables)
```
Comment (rousskov) [inline src/cf.data.pre]: I would rephrase this sentence to polish and clarify: `Use the mask to preserve the mark previously set by other agents (e.g., iptables)` or `The mask may be used to preserve the mark previously set by other agents (e.g., iptables)`. The use of the mask is conditional on admin needs/desires, not on the "other agents" actions.

File: src/client_side_request.cc [hunk 4]
```diff
@@ -1794,9 +1794,14 @@ ClientHttpRequest::doCallouts()
...
+                // Clear the bits given by the mask in existing CONNMARK
+                // and XOR it with a new mark.
+                nfmark_t mark = getConn()->clientConnection->nfmark & ~(l->nfmask);
+                mark ^= l->nfmark;
```
Comment (rousskov) [inline src/client_side_request.cc]: I have two concerns here:

1. To an non-expert like me, it seems that the configured mask should apply to the configured `nfmark` value as well.
2. If `nfmask` is applied to `nfmark`, then the XOR part seems unnecessary complex to me. The `~nfmask` clearing above guarantees that all the relevant bits in `mark` are going to be zero. Similarly, all the irrelevant bits in `nfmark` will be zero due to (1). Thus, applying a simpler OR logic is sufficient: `0 xor x == 0 or x`

In summary, I wonder whether you want an OR operation with a mask (instead of XOR without a mask):

```
mark |= (l->nfmark & l->nfmask);
```

This is not my area of expertise so if XOR-without-a-mask is how we are supposed to implement this to accommodate some unusual/advanced use cases, please disregard the above concern, with my apologies. Otherwise, please also adjust the documentation to match both changes discussed above.

File: src/cf.data.pre [hunk 5]
```diff
@@ -2570,7 +2570,15 @@ DOC_START
...
+	should be preserved.
+
+	The directive clears the original connmark's bits given by the mask and XORes
+	it with a new mark. The result is set as a new connmark.
+	The default mask is 0xffffffff, which means that original mark will be overwritten.
```
Comment (rousskov) [inline src/cf.data.pre]: I would rephrase to simplify and avoid dependency on a hard-coded mask value: `If no mask is specified, then Squid overwrites the entire original mark with the configured mark.`

**Comment by rousskov** [comment]
I polished the title to make it just a tad simpler and to remove the trailing period that kind of clashes with the automatically appended PR number when the PR is committed. Please feel free to adjust to your liking, including reverting my changes.

To minimize noise, the PR description (i.e., future commit message body) should probably be polished _after_ the XOR/masking concern is addressed.

File: src/client_side_request.cc (hunk 4 above)
Comment (gozzy) [reply src/client_side_request.cc]: I've borrowed the logic from iptables' 'set-xmark' option that works as described here: https://stackoverflow.com/a/22003527

Probably the documentation should be adjusted to cover all possible cases.

Comment (rousskov) [reply src/client_side_request.cc]: Aha, I think I know where the core of the problem is. We need to make a choice between two very different marking semantics. You picked complex/adjustment-focused `--set-xmark` semantics over the simple/assignment-focused `--set-mark` semantics. If you look at the actual [documentation](http://ipset.netfilter.org/iptables-extensions.man.html) that the Stackflow article is quoting, you will see both operations defined:

```
--set-xmark value[/mask]
    Zero out the bits given by mask and XOR value into the ctmark. 

The following mnemonics are available for --set-xmark:

--and-mark bits
    Binary AND the ctmark with bits. (Mnemonic for --set-xmark 0/invbits, where invbits is the binary negation of bits.) 
--or-mark bits
    Binary OR the ctmark with bits. (Mnemonic for --set-xmark bits/bits.) 
--xor-mark bits
    Binary XOR the ctmark with bits. (Mnemonic for --set-xmark bits/0.) 
--set-mark value[/mask]
    Set the connection mark. If a mask is specified then only those bits set in the mask are modified. 
```

As I expected, the set-xmark semantics is very powerful -- it can do a lot of fancy operations. However, AFAICT, it cannot fully implement the set-mark semantics! Notice how the mnemonics section of the above documentation gives equivalent expressions for all other options _except_ `--set-mark`? The iptables documentation misleads -- the set-mark option is not a mnemonic for a subset of set-xmark use cases. It is a stand-alone option!

Now that we understand that there are two different operations, we have to pick the one we should implement in Squid. Before this PR, Squid supported both operations because when the mask is all-ones, there is no difference between the two. However, there is a difference going forward when the masks become non-trivial.

* If we implement just the set-xmark semantics, then the admin will not be able to tell Squid to set the last byte of the mark to `0xB` without touching anything else.
* If we implement just the set-mark semantics, then the admin will not be able to tell Squid, for example, to add  `0xB` bits to the existing mask bits without turning any existing bits off. There are other mask adjustment operations that the admin will not be able to do either.

The Squid directive is _not_ named `clientside_xmark` -- there is no visual warning for the admin that something complex is going on.

Given all of the above considerations/concerns, we have two reasonable choices IMHO:

1. Implement just the set-mark algorithm. This option covers only the simple marking operations. It is easy to adjust your PR to implement this option.
2. Implement both set-xmark and set-mark algorithms with some syntax sugar (or a new directive) to distinguish the two. This option covers both complex and simple marking operations. Its implementation requires significant additional work.

If _you_ need the set-xmark power, then I suggest moving towards option 2. Otherwise, we can start with option 1 for now and add set-xmark support later when somebody needs it. This is just my non-expert opinion. Others may suggest a different decision-making algorithm (and different implementation options).

What do _you_ think? And do you want to wait for other reviews before modifying the PR further?

Comment (gozzy) [reply src/client_side_request.cc]: Yes, it seems that I've tried to do it the hard way. Currently I think that set-mark logic would be a better and more obvious option.

Comment (gozzy) [reply src/client_side_request.cc]: I'll try to fix current issues (and change the logic, of course) and then we can get back to review.

File: src/cache_cf.cc (hunk 1 above)
Comment (yadij) [reply src/cache_cf.cc]: Also, be aware that CONNMARK are 32-bit values not 8-bit so the %02X is plain wrong anyway.

**Comment by gozzy** [comment]
I hope it's a bit better now.

File: src/cf.data.pre [hunk 6]
```diff
@@ -2570,7 +2570,16 @@ DOC_START
 	Allows you to apply a Netfilter mark value to packets being transmitted
 	on the client-side, based on an ACL.
 
-	clientside_mark mark-value [!]aclname ...
+	clientside_mark mark[/mask] [!]aclname ...
+
+	Mark and mask are unsigned integers (hex, octal, or decimal).
+	The mask may be used to preserve the mark previously set by other agents
```
Comment (rousskov) [inline src/cf.data.pre]: s/the mark/marking/
or
s/the mark/portions of the mark/
or
s/the mark/mark bits/

File: src/cf.data.pre [hunk 7]
```diff
@@ -2570,7 +2570,16 @@ DOC_START
...
+	(e.g., iptables).
+
+	The directive clears the original connmark's bits given by the mask and ORs
+	it with a new mark. The result is set as a new connmark.
```
Comment (rousskov) [inline src/cf.data.pre]: The first sentence is too complex (because it is trying to translate code into English) and also a little buggy (we do not OR with the _whole_ new mask). I suggest simplifying by focusing on the code result (not the process) and assuming that the two sentences will be interpreted together (instead of trying to talk about the mask effect in each sentence). Something like this may work better (shown without formatting):

> A matching rule replaces the original connmark with the configured one. If a mask is specified, then only the masked bits are affected.

I also suggest adding a paragraph to document handling of multiple rules:

> The directive rules are evaluated in the configuration order. The first matching rule wins.

File: src/client_side_request.cc [hunk 8]
```diff
@@ -1794,9 +1794,13 @@ ClientHttpRequest::doCallouts()
             ACLFilledChecklist ch(NULL, request, NULL);
             ch.src_addr = request->client_addr;
             ch.my_addr = request->my_addr;
-            nfmark_t mark = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch);
-            if (mark)
+            if (const auto mc = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch)) {
+                // Clear the bits given by the mask in existing CONNMARK
+                // and OR it with a new mark.
+                nfmark_t mark = getConn()->clientConnection->nfmark & ~(mc->nfmask);
+                mark |= mc->nfmark;
```
Comment (rousskov) [inline src/client_side_request.cc]: Again, we muast apply nfmask to nfmark. Use mc->maskedMark() discussed elsewhere.

File: src/ip/QosConfig.h [hunk 9]
```diff
@@ -43,12 +43,13 @@ class acl_nfmark
...
 
 public:
-    acl_nfmark() : next(NULL), aclList(NULL), nfmark(0) {}
+    acl_nfmark() : next(NULL), aclList(NULL), nfmark(0), nfmask(0xffffffff) {}
     ~acl_nfmark();
 
     acl_nfmark *next;
     ACLList *aclList;
     nfmark_t nfmark;
+    nfmark_t nfmask;
```
Comment (rousskov) [inline src/ip/QosConfig.h]: Please use a new `ConnMarkConfig nfMark` field instead of these two low-level fields.

File: src/client_side_request.cc (hunk 8 above)
Comment (gozzy) [reply src/client_side_request.cc]: AFAIK, set-mark in iptables works like this: ctmark = (ctmark AND NOT mask) OR value.

File: src/acl/ConnMark.cc [hunk 10]
```diff
@@ -77,7 +44,7 @@ Acl::ConnMark::match(ACLChecklist *cl)
     const auto connmark = checklist->conn()->clientConnection->nfmark;
 
     for (const auto &m : marks) {
-        if ((connmark & m.second) == m.first) {
+        if ((connmark & m.nfmask) == m.nfmark) {
```
[the remark this answers is not in the document]
Comment (gozzy) [reply src/acl/ConnMark.cc]: I am not sure about it. Iptables' man states that a mask is applied to ctmark before comparison (if I understood the man correctly)

File: src/client_side_request.cc (hunk 8 above)
Comment (rousskov) [reply src/client_side_request.cc]: Documentation says otherwise: `If a mask is specified then only those bits set in the mask are modified`. However, the documentation may lie. If you can test what iptables do, please test.

File: src/acl/ConnMark.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/acl/ConnMark.cc]: What you are saying (the mask is applied on LHS) does not contradict what I am saying (the mask should be applied on LHS and on RHS). If you can test what iptables actually do, please test. For example, use an 0xFF mark with a 0xF mask and see whether the masked-out bits matter.

File: src/client_side_request.cc (hunk 8 above)
Comment (gozzy) [reply src/client_side_request.cc]: According to [this]https://github.com/theojulienne/iptables/blob/master/extensions/libxt_MARK.c#L126 and [this]https://github.com/torvalds/linux/blob/9cb9c07d6b0c5fd97d83b8ab14d7e308ba4b612f/net/netfilter/xt_connmark.c#L92, the mask is applied to original mark. Also, if I understand correctly, netfilter expresses set-mark through XOR.

Comment (gozzy) [reply src/client_side_request.cc]: Actually, iptables has very odd documentation :) The statement can be also treated as 'if we have a mask, we zero out corresponding bits in a source mark and then apply a new mark to it'.

Comment (gozzy) [reply src/client_side_request.cc]: We've done some tests when clientside_mark ACL was added and everything worked as expected.

Comment (gozzy) [reply src/client_side_request.cc]: However, we can run additional tests on monday, if needed. Any more remarks on this PR?

Comment (rousskov) [reply src/client_side_request.cc]: Sorry for not being clear: The tests I mentioned were concerning iptables, not Squid. I understand that your Squid implementation matches your expectations, and that is great, but that is not the issue here. The iptables questions are:

1. Does `--mark 0xBF/0xF` match an original/preset 0xAF mark?
2. Given no/zero original/preset mark, does `--set-mark 0xFF/0xF` set a 0xF mark or a 0xFF mark?

I do not know the exact test commands for figuring out the above, but I assume that you do.

> Docs: If a mask is specified then only those bits set in the mask are modified

> The statement can be also treated as 'if we have a mask, we zero out corresponding bits in a source mark and then apply a new mark to it'.

I disagree that it is a reasonable interpretation because the docs talk about restricting _modification _ and modification only makes sense if applied to the connection mark, not any configured value (which is a _constant_ in this context). However, again, iptables documentation may be lying or be misleading -- we should not rely on it (alone).

BTW, I would agree with your interpretation if you replace "it" with "them" or "those bits".

In fact, even relying on the iptables behavior/code _alone_ is dangerous -- it could be buggy when it comes to unusual configurations such as 0xBF/0xF. There should be some common, well-known approach to these issues that all tools, including iptables and Squid should follow instead. Unfortunately, I do not know about that approach because I am not a packet filtering expert. I assume that you do not know it either (because you have not referred to it yet). And nobody else has chimed in. Thus, we are trying to make progress by testing iptables and looking at iptables code and documentation. This is not the best or an efficient way to do it, but it may still result in a well-understood (if not necessarily correct) commit.

> Any more remarks on this PR?

I do not know yet. If you are done with addressing the previous comments, then I will review the changes once again after this 0xBF/0xF issue is fully addressed.

Comment (gozzy) [reply src/client_side_request.cc]: A quick test shows:
--mark 0xBF/0xF does not match 0xAF
--set-mark 0xFF/0xF sets 0xFF

The test rules were:
```
iptables -t nat -A OUTPUT -p tcp --dport 80 -j MARK --set-mark 0xAFk
iptables -A OUTPUT -m mark 0xBF/0xF -j DROP
```

Comment (gozzy) [reply src/client_side_request.cc]: I've addressed all previous comments in current version (except of a marking logic that is currently beeing discussed).

Comment (rousskov) [reply src/client_side_request.cc]: @hno, I noticed you authored MARA's connection marking Linux [code](https://github.com/torvalds/linux/blob/9cb9c07d6b0c5fd97d83b8ab14d7e308ba4b612f/net/netfilter/xt_connmark.c#L30). This PR adds `/mask` support to Squid's existing [clientside_mark](http://www.squid-cache.org/Doc/config/clientside_mark/) directive. Assuming that we are implementing an iptables-like `--set-mark` semantics (and _not_ `--set-xmark` semantics!), which of the following options for treating "excessive" mark values (as in 0xAB/0xF) would you recommend?

1. Excessive bits should be ignored: Setting 0xAB/0xF means setting 0xB/0xF.
2. Excessive bits should be honored: Setting 0xAB/0xF means setting 0xAB/0xFF.
3. Excessive bits is a configuration error. Different reasonable people may expect different outcomes when setting 0xAB/0xF.

Same questions apply to matching 0xAB/0xF when using a clientside_mark ACL, but I assume the answers to the "matching" questions would be the same as the answers to the "setting" questions above.

Comment (rousskov) [reply src/client_side_request.cc]: @gozzy, thank you for testing those cases! I believe both tests imply that 0xAB/0xF means 0xAB/0xFF to iptables (which is what your latest code implements). Let's continue to use that semantics at least until an expert chimes in. Please ignore my maskedValue() suggestions. 

I will check the latest code shortly.

File: src/cf.data.pre [hunk 11]
```diff
@@ -2570,7 +2570,16 @@ DOC_START
...
 
-	clientside_mark mark-value [!]aclname ...
+	clientside_mark mark[/mask] [!]aclname ...
+
+	Mark and mask are unsigned integers (hex, octal, or decimal).
+	The mask may be used to preserve marking previously set by other agents
+	(e.g., iptables).
+
+	The directive clears the original connmark's bits given by the mask and ORs
+	it with a new mark. The result is set as a new connmark. I.e. if a mask is
```
Comment (rousskov) [inline src/cf.data.pre]: We must polish this paragraph for clarity and to avoid the questions we had about similar iptables documentation. How about the following:

> A matching rule replaces the original connmark value with the configured mark. If a mask is also specified, then the masked bits of the original value are zeroed, and the configured mark is ORed with that adjusted value. For example, setting a 0xAB/0xF mark for a connection with a 0x5F connmark, results in a 0xFB connmark (rather than a 0xAB or 0x5B connmark).

It may be useful to also add the following paragraph:

> This directive semantics is similar to iptables --set-mark rather than --set-xmark functionality.

I also suggest adding a paragraph to document handling of multiple rules:

> The directive rules are evaluated in the configuration order. The first matching rule wins.

**Comment by rousskov** [comment]
@gozzy, AFAICT, `src/ip/QosConfig.cc:540` has a stale `nfmark` that needs updating.

**Comment by gozzy** [comment]
Done, updated the PR.

**Comment by rousskov** [comment]
OK to test.

**Comment by gozzy** [comment]
Thank you for the review! Regarding matches() - you mean adding a description to a header file? Something like "checks if given mark matches cofigured mark"?

**Comment by rousskov** [comment]
Yes, see https://github.com/squid-cache/squid/pull/166#discussion_r171994283 for the suggested text. Adding explicit documentation is not important in this case because the inlined code speaks for itself IMO. That is why I am not requesting another dedicated revision.

N.B. I adjusted the PR description (future commit message body) to match the current state of the code and to fix formatting. Please adjust further as needed.

**Comment by gozzy** [comment]
Added the comment and adjusted the commit message (just to not forget it).

**Comment by rousskov** [comment]
FYI: You do not need to adjust individual branch commit messages to match the future master commit message because, by default, we are doing squash-merge and using the PR title/description as the master commit message. This approach allows us to edit the future master commit message as needed and to avoid in-branch squashing of individual commits as the branch gets reviewed/polished.

**Comment by rousskov** [comment]
@yadij, I plan to commit this PR after it passes the tests, but since connection marking is not my area of expertise, the PR can especially benefit from a third pair of eyes.

**Comment by yadij** [comment]
Just to clarify for me, this is following iptables behaviour for --set-mark right?

**Comment by rousskov** [comment]
Thank you for fixing typos in the PR description.

**Comment by gozzy** [comment]
Slightly updated description and comments (clientside_mark directive sets a packet mark, not connmark).

**Comment by gozzy** [comment]
Well, hindsight is 20/20... As clientside_mark directive works with packet marks, this makes masking useless... Unless someone wants to combine incoming connmark with something and set it as a packet mark. Furthermore, there will be an oddity: clientside_mark directive sets packet marks, but the ACL of the same name works with connmarks. Frankly speaking, despite the commit neither does anything bad nor breaks the default behavior, it's not worthwhile (IMHO).

**Comment by rousskov** [comment]
> clientside_mark directive works with packet marks, [making] masking useless

For my own education sake, why is packet masking useless? Iptables' MARK module supports masking for packets AFAICT so somebody must find it valuable, I guess. Why do you think it is useless?

> there will be an oddity: clientside_mark directive sets packet marks, but the ACL of the same name works with connmarks

Ouch! That sounds like a bug to me. This PR did _not_ introduce that bug (it only exposed it by fixing the old directive documentation), but we should decide (here or elsewhere) whether to 

1. enhance the newer ACL to work with packets as well as connections
2. enhance the older directive to work with connections as well as packets

**Comment by rousskov** [comment]
@gozzy, squashing your branch changes makes it difficult to know what has changed compared to the previous branch version. Please do not squash (by default).

**Comment by gozzy** [comment]
> For my own education sake, why is packet masking useless?

It can be useful in a common case (e.g., when packet passes through different iptables rules). However in this case it seems useless to me because the mark is applied to a packets in "squid -> client" connection (if I understand SO_MARK socket option correctly). Squid has no predecessors here. Perhaps I get the things wrong, but the manual states that SO_MARK affects packets that are **_sent_** through the socket.

**Comment by gozzy** [comment]
I suppose that enhancing clientside_mark to make It able to work with connections may do a trick. This request further investigation.

**Comment by rousskov** [comment]
> the mark is applied to a packets in "squid -> client" connection. Squid has no predecessors here

I see your point. Thank you for explaining this to me.

It is my understanding that you were working on this PR to solve a real-world problem. This PR (as it is now) does not solve that problem. Applying the configured mark/mask to the _connection_ rather than the packet would solve the problem that you need to solve. Is this summary accurate?

**Comment by yadij** [comment]
@gozzy, the qos_flows directive sets outgoing marks as well and controls how the server marking (if any) is relayed on MISS traffic. Having a mask on this as well will help reduce the problem of this directive overwriting the preserved server mark.

**Comment by gozzy** [comment]
@rousskov, you are right, we've tried to solve a problem with native FTP relay and connmark-based iptables rules. In general this PR solves the problem, but this mixes up connection and packet marks. This is not good and (possibly) may lead to other problems. Maybe this PR is not as bad as it may seem to me now, because there is a possiblity to turn a packet mark into a connection mark.

@yadij, does the QoS-related thing require any additional work?

**Comment by yadij** [comment]
> does the QoS-related thing require any additional work?

Not that I am aware of beyond what you have done already.

**Comment by gozzy** [comment]
In this case we could give it a green light. I haven't thought about QoS at the beginning,  but if the PR may potentially help with it - why not?

**Comment by rousskov** [comment]
> we've tried to solve a problem with native FTP relay and connmark-based iptables rules. In general this PR solves the problem, but this mixes up connection and packet marks.

You lost me here. How can _this PR_ solve any marking problem if masking Squid-originated packets is pointless because there is no "previous" mark value to preserve? Why do you need to specify a mask with clientsode_mark? I must be missing something important. AFAICT, besides masking in clientsode_mark context, this PR does not add any other functionality; it fixes documentation and improves code quite a bit, but those are not functionality improvements...

**Comment by rousskov** [comment]
> Amos: Having a mask on clientside_mark directive will help reduce the problem of it overwriting the preserved server mark from qos_flows.

> squid.conf.documented: Any mark values set [by clientside_mark] will be overwritten by mark values in qos_flows.

Both statements cannot be true AFAICT. Either qos_flows is applied first (then Amos is right, documentation lies, and the functionality changes in this PR are useful) or clientside_mark is applied first (then the documentation is right, and the functionality changes in this PR are harmful).

**Comment by gozzy** [comment]
@rousskov, in short, this PR allows to take a connection mark set by iptables (before squid intercepted a connection) and set an adjusted result as a packet mark (in squid to client direction). In some cases (like ours) this could be useful, but for most users this might be hard to understand (especially when two things with similar names behave differently).

The root of all evil is that Squid treats one mark in two different ways. It gets a connection mark on one end and turns it into packet mark on the other end.

PS: the documentation itself if fine, QoS mark is set after clientside_mark. I've just checked this out. This doesn't make the PR look harmful because everything works as it did earlier.

**Comment by gozzy** [comment]
Maybe this still can be merged, but before that we need to change clientside_mark description according to what is written above.

**Comment by rousskov** [comment]
> The root of all evil is that Squid treats marks in two different ways. It gets a connection mark on one end and turns it into packet mark on the other end. I don't know whether it's right or wrong, it's what it is.

Agreed. IMHO, it is wrong because it is non-standard, counter-intuitive, and asymmetrical compared to the ACL with the same name. AFAICT, you and the ACL authors got caught in this trap. I suspect many admins suffer the same fate and just brute-force their way through it while silently pummeling Squid. Stable Squids may need to preserve this functionality for backward compatibility reasons, but enhancing this broken functionality further by adding masking support would only make it worse -- we would have more (broken) configurations to worry about when we finally decide to fix this mess.

We could stop this madness now, in this PR, by, for example, adding a clientside_mark parameter to distinguish between packet and connection markings (and adding connection marking support). Or we could just rename the clientside_mark directive and/or the ACL to emphasize which one is packet-based and which is connmark-based. However, somebody other than me should lead this effort -- this is not my area of the expertise, and I already spent way too much time on this PR!

**Comment by gozzy** [comment]
I'm not giving up yet. Probably this will result in a new PR that will try to stop the mess with marks, or at least reduce it.

**Comment by rousskov** [comment]
Glad to hear that! To minimize rewrites, I recommend posting the proposed cf.data.pre changes (or equivalent; with an accompanying justification text) _before_ modifying any code. This can be done as a Work-in-Progress PR by prefixing the PR title with `WIP: `.
